Adds support for sdl_image and sdl_ttf (also other libs)

WizzardMaker opened the pull request:
I needed the new versions of SDL3_image and SDL3_ttf, and I liked the implementation of this library for SDL3.
I've refactored the bindings script to be able to use different external libraries.

For now only image and ttf are added, as they are the only ones I need, but the code change allows for easy addition of the other SDL libs.

smoogipoo commented:
I think we should separate the csprojs, so we'd have the current `SDL3-CS` one and add `SDL3-CS.TTF`, etc. Just to reduce the size of native dependencies for consuming projects (for example osu! will likely never use TTF).

smoogipoo commented on SDL3-CS/SDL3/ClangSharp/SDL_audio.g.cs:
```diff
@@ -28,6 +28,7 @@ 3. This notice may not be removed or altered from any source distribution.
 
 namespace SDL
 {
+    [NativeTypeName("int")]
```
Did you generate bindings manually? If so, please use the following in the future (will add to readme):

```sh
docker build -t 'sdl-gen' .
docker run --rm -v .:/app -w /app -it sdl-gen
```

WizzardMaker commented:
> I think we should separate the csprojs, so we'd have the current `SDL3-CS` one and add `SDL3-CS.TTF`, etc. Just to reduce the size of native dependencies for consuming projects (for example osu! will likely never use TTF).

Technically, dotnet will only lazy load native dependencies when those functions are first called. So without calling any SDL3.TTF_* no ttf dll would be required.
But I do see the benefits of separations of concerns here. It would also make it easier for developers when _ttf or _image related functions are in their own space.
But I think separating them into different csproj would be a bit overkill - shouldn't different classes (e.g. `SDL3_ttf`) be sufficient?

smoogipoo commented:
I meant the inclusion of the native libs, because we're also going to have to build + package them here. The native `SDL2_ttf.dll` for instance is a 1.8MB dependency, which will be included in the output regardless of whether or not it's used.

WizzardMaker commented:
Ah, yes of course. I haven't thought about the nuget package implications

I can split it up into their respective csproj with PackageReferences to the SDL3-CS package

smoogipoo commented:
For Android, I'm tempted to multi target and require the workflow for build. Including (and removing) the existing package (@Susko3 would you agree with that?).

Susko3 commented:
I agree with multi targetting Android. The `android` workload probably won't be required if running the `dotnet build` command for a specific RID (`--runtime`), so that's not really an issue.

Susko3 commented on SDL3-CS/SDL3_ttf/ClangSharp/SDL_ttf.g.cs:
```diff
@@ -0,0 +1,506 @@
+        public const int TTF_SUBSTRING_LINE_END = 0x00000004;
+
+        [NativeTypeName("#define TTF_SUBSTRING_TEXT_END 0x00000008")]
+        public const int TTF_SUBSTRING_TEXT_END = 0x00000008;
```
Ditto for typedef'd enum.

WizzardMaker commented:
I've now split up the projects into their respective libraries.
The native build script was also updated to the best of my abilities without running the pipeline or access to macOS/linux

There is currently a chicken or egg problem, where the sister projects depend on the main project nuget package, but before they can work they need a change in the main project (see af476b4cc4357abdfb2a9e1ec46677addf2f1c0e)

smoogipoo commented on External/build.sh:
```diff
@@ -102,4 +102,48 @@ elif [[ $RUNNER_OS == 'macOS' ]]; then
+    cp SDL_image/install_output/bin/tiff.dll ../native/$NAME/tiff.dll
+elif [[ $RUNNER_OS == 'Linux' ]]; then
+    cp SDL_image/install_output/lib/libSDL3_image.so ../native/$NAME/libSDL3_image.so
+    # TODO: find out if webp, etc. are also needed on linux here
```
With these TODOs, is it just that you haven't tested if these are output at all, or only if they're necessary at runtime? If they're output they I think we should include them.

WizzardMaker replied:
Output at all. I know that the Linux build statically compiles a lot more dependencies than a windows build by default

I agree that they should be included, if they are built

smoogipoo commented:
Have applied a few fixes here. Of importance (to me, for testing), I replaced the `PrivateAssets` pkgref to SDL3-CS with a project reference.

Haven't tested things yet but have checked the SG output.

WizzardMaker commented:
I have used a PackageReference to SDL3_CS for dependency management when included in another project via nuget. Otherwise with just a project reference the SDL3-CS binary would either need to be manually added as dependency in nuget or be included in the bundle as well. Bundling shouldn't be done, as that would lead to multiple copies when more than one extra library is included

For that Package reference to work we either need to first publish a version with the InternalsVisibleTo change, or add the -t:package output of SDL3-CS as nuget store source to the other projects

smoogipoo commented:
I'd really like to avoid local nuget package sources. We've had to do that in other projects and it gets really finicky if you don't version it appropriately (i.e. timestamp or commit _depth_).

In theory it should be guided by the generated `.nuspec`, which looks like this:
```
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2013/05/nuspec.xsd">
  <metadata>
    <id>ppy.SDL3_ttf-CS</id>
    <version>1.0.0</version>
    <authors>ppy Pty Ltd</authors>
    <license type="expression">MIT</license>
    <licenseUrl>https://licenses.nuget.org/MIT</licenseUrl>
    <projectUrl>https://github.com/ppy/SDL3-CS</projectUrl>
    <description>Package Description</description>
    <releaseNotes>Automated release.</releaseNotes>
    <copyright>Copyright (c) 2024 ppy Pty Ltd</copyright>
    <repository type="git" url="https://github.com/ppy/SDL3-CS" commit="b3dec6b80cec964abdb7143fdbc89af1c6ee8b5d" />
    <dependencies>
      <group targetFramework="net8.0">
        <dependency id="ppy.SDL3-CS" version="1.0.0" exclude="Build,Analyzers" />
      </group>
    </dependencies>
  </metadata>
</package>
```

I'll need to read more on the documentation, as I find it hard to believe `PackageReference` is the way this should be done.

GRATHRRAM commented:
Can you marge It. Please

smoogipoo commented:
> Is there anything the community could do to finalize the support for sdl_image and sdl_ttf?

The SDL_image (and presumably SDL_ttf) native libs currently fail to build, see errors in https://github.com/smoogipoo/SDL3-CS/actions/runs/15130408196

I've tried to fix it real quick to no avail, so I'm dropping this for the time being. If you manage to fix it then I can try fast-forwarding this PR.

smoogipoo commented:
@jairov4 Thanks. Have pushed here just to simplify things.

I've also made a few changes myself around the deploy process. I'll attempt a deploy as is, but note that it's not going to have ios/android libs for now - those aren't built by the GHA script:

https://github.com/ppy/SDL3-CS/blob/e34ff3f25a7d29f6a79f1c61aab4064155cd303e/.github/workflows/build.yml#L142-L149

https://github.com/ppy/SDL3-CS/blob/e34ff3f25a7d29f6a79f1c61aab4064155cd303e/.github/workflows/build.yml#L87-L88

smoogipoo commented:
Hi all, sorry for the wait on this. Packages are now published:
- https://www.nuget.org/packages/ppy.SDL3-CS/2025.620.0
- https://www.nuget.org/packages/ppy.SDL3_ttf-CS/2025.620.0
- https://www.nuget.org/packages/ppy.SDL3_image-CS/2025.620.0
